In [ ]:
import numpy as np
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split



In [ ]:

# -------------------------
# データ数切り替えフラグ
# -------------------------
#USE_MORE_DATA = False   # ← False にすると「データ増加前」のケース
USE_MORE_DATA = True    # ← True にすると「データ増加後」のケース
# -------------------------


In [ ]:

np.random.seed(42)

# 実行関数
def f(x):
    return np.sin(2 * np.pi * x)

# データ数をフラグで変更
if USE_MORE_DATA:
    n_samples = 1000                                    # データ増加（過学習改善を期待）
    print("データ数: 多い（", n_samples, "）")
else:
    n_samples = 10                                      # 少ないデータ（過学習を誘発）
    print("データ数: 少ない（", n_samples, "）")

# データ生成
X = np.random.rand(n_samples, 1)                        # [0,1) の一様乱数
y = f(X).ravel() + 0.1 * np.random.randn(n_samples)     # ノイズ付き

# 訓練・テスト分割
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

# 高次多項式特徴（次数 15）
degree = 15
poly = PolynomialFeatures(degree=degree)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# 線形回帰（多項式回帰）
model = LinearRegression()
model.fit(X_train_poly, y_train)

# 予測
y_train_pred = model.predict(X_train_poly)
y_test_pred  = model.predict(X_test_poly)

# 誤差
train_mse = mean_squared_error(y_train, y_train_pred)
test_mse  = mean_squared_error(y_test, y_test_pred)

print("次数:", degree)
print(f"訓練   MSE: {train_mse:.6f}")
print(f"テスト MSE: {test_mse:.6f}")


データ数: 多い（ 1000 ）
次数: 15
訓練   MSE: 0.009782
テスト MSE: 0.009357


**● データ数: 少ない（ 10 ）**  
　次数: 15  
　訓練   MSE:  0.000000  
　テスト MSE: 44.030760  


  
**●データ数: 多い（ 1000 ）**  
　次数: 15  
　訓練   MSE: 0.009782  
　テスト MSE: 0.009357  

データが少ないときは、訓練誤差が極端に小さいのに対しテスト誤差が非常に大きく、両者の乖離が大きいため、モデルが訓練データに過度に適合しており過学習が発生していると判断できる。  
このとき用いている MSE（Mean Squared Error：平均二乗誤差） は、予測値と実際の値の差を二乗して平均した指標で、値が小さいほど予測が実際に近いことを意味する。  
訓練 MSE が極端に小さいのにテスト MSE が大きい場合、モデルが訓練データの細部やノイズまで記憶してしまい、未知データでは誤差が大きくなるため、典型的な過学習の状態といえる。
